server/conn.go

```diff
@@ -147,8 +147,11 @@ func (cc *clientConn) writeInitialHandshake() error {
 	data = append(data, 0)
 	// capability flag lower 2 bytes, using default capability here
 	data = append(data, byte(defaultCapability), byte(defaultCapability>>8))
-	// charset, utf-8 default
-	data = append(data, uint8(mysql.DefaultCollationID))
+	// charset
+	if cc.collation == 0 {
```

Comment:
yep, but `TestInitialHandshake` in `conn_test.go` initialized a `clientConn` without collation field, and then called `cc.writeInitialHandshake()`, without this branch the case would fail.